storage/remote/client.go

```diff
@@ -199,25 +205,77 @@ type RecoverableError struct {
 	retryAfter model.Duration
 }
 
+// Attempt a HEAD request against a remote write endpoint to see what it supports.
+func (c *Client) GetProtoVersions(ctx context.Context) (string, error) {
+	// If we are in Version1 mode then don't even bother
+	if c.rwFormat == Version1 {
+		return RemoteWriteVersion1HeaderValue, nil
+	}
+
+	httpReq, err := http.NewRequest("HEAD", c.urlString, nil)
+	if err != nil {
+		// Errors from NewRequest are from unparsable URLs, so are not
+		// recoverable.
+		return "", err
+	}
+
+	// Set the version header to be nice
+	httpReq.Header.Set(RemoteWriteVersionHeader, RemoteWriteVersion20HeaderValue)
+	httpReq.Header.Set("User-Agent", UserAgent)
+
+	ctx, cancel := context.WithTimeout(ctx, c.timeout)
+	defer cancel()
+
+	httpResp, err := c.Client.Do(httpReq.WithContext(ctx))
+	if err != nil {
+		// We don't attempt a retry here
```

Comment:
What does this mean? That we retry somewhere else? We never retry? Let's be specific (or drop the comment, it might not help us). Let's drop it I would say (:

Reply:
Right now we're actually not doing anything with the return values from this function. We should IMO. If the user has configured their prometheus to use rw2.0 but the reciever doesn't respond correctly we should likely fail fast and have prometheus exit, or at least include a rw config option that allows them to choose to fallback to 1.0.